### ISSUE TITLE
Dispatch StorageEvent events to other windows when modifying storage instances

### ISSUE THREAD

**Issue by jdm** [servo/servo#5196]
Spec: https://html.spec.whatwg.org/multipage/webstorage.html#the-storageevent-interface, https://html.spec.whatwg.org/multipage/webstorage.html#send-a-storage-notification, https://html.spec.whatwg.org/multipage/webstorage.html#the-sessionstorage-attribute:send-a-storage-notification, https://html.spec.whatwg.org/multipage/webstorage.html#the-localstorage-attribute:send-a-storage-notification
Code: http://mxr.mozilla.org/servo/source/components/script/dom/storage.rs

The bit about "for every Document object whose Window object's localStorage attribute's Storage object is associated with the same storage area, other than x, send a storage notification" is the important bit to figure out here.

**Comment by KiChjang** [comment]
I just read through the spec, and there's a bit under `sessionStorage` that says "Each top-level browsing context has a unique set of session storage areas, one for each origin."

So if we can figure out `Storage` object *x*'s origin, we can loop through every `sessionStorage` object and check whether its `Storage` object has the same origin as *x*, and if so, send a storage notification to that `Storage` object.

**Comment by zslayton** [comment]
I'm going to be traveling for the next week or so, but if this is still unclaimed when I return I'd love to work on it.

**Comment by snf** [comment]
I'm working on this. I hope can submit the EventStorage interface implementation later.

**Comment by snf** [comment]
I have almost everything implemented but I'm not sure how to iterate on Document/Windows. I'm already executing inside script_task::MainThreadRunnable so I have access to ScriptTask, but page and root_page() are private and I can't find any other method that would help me here.
I have been thinking in making root_page() public but you may have any other idea on how to implement it.

**Comment by jdm** [comment]
Go ahead and make it public.

**Comment by snf** [comment]
About the remaining tests for Storage:

<jga_> I'm working in Storage and I see that some of the tests use the notation localStorage["1"] or localStorage["key"], which is the actual function being called?
<jga_> getItem or key?
<jga_> or none and that's why I'm failing and I have to implement another trait
<jga_> I guess there is lacking something that connects the javascript representation to the actual functions

**Comment by jdm** [comment]
That should be going through the getter method in the WebIDL, which corresponds to NamedGetter/NamedSetter/NamedCreator, depending on the expression.

**Comment by jdm** [comment]
However, #5223 might be causing this to mess up.

**Comment by snf** [comment]
I'm investigating it a little more and it seems that storage_task is not updating the values. Will keep working on it tomorrow.

**Comment by snf** [comment]
Ok, it seems that NamedGetter/Setter is being cached somewhere. If I update it with the NamedSetter, the NamedGetter and GetItem returns the right value, but if I update with SetItem, only GetItem returns the right value and NamedGetter returns the last value set with NamedSetter.

Hope that makes sense somehow.

**Comment by Ms2ger** [comment]
There is a bug in the generated code: #5223.

**Comment by snf** [comment]
Right, wasn't reading #5223 very well.

**Comment by asajeffrey** [comment]
IRC chat with @jdm: http://logs.glob.uno/?c=mozilla%23servo&s=1+Nov+2016&e=1+Nov+2016#c552341

TL;DR the notifications should go via the constellation, since we need to target every same-origin window, no matter which script thread they are in.
